examples/reshare/src/orchestrator/actor.rs

```diff
@@ -164,51 +175,108 @@ where
-                        .enter_epoch(
-                            transition.epoch,
-                            transition.poly,
-                            transition.share,
-                            transition.participants,
-                            &mut pending_mux,
-                            &mut recovered_mux,
-                            &mut resolver_mux,
-                        )
-                        .await;
-                    engines.insert(transition.epoch, engine);
+                    // Fetch the finalization certificate for the last block within the subchannel's epoch.
+                    // If the node is state synced, marshal may not have the finalization locally, and the
+                    // peer will need to fetch it from another node on the network.
+                    let boundary_height = last_block_in_epoch(BLOCKS_PER_EPOCH, epoch as u64);
+                    let Some(finalization) = self.marshal.get_finalization(boundary_height).await else {
+                        debug!(epoch, ?from, "missing finalization for old epoch");
+                        continue;
+                    };
+
+                    warn!(
+                        epoch,
+                        boundary_height,
+                        ?from,
+                        "received message on recovery/pending network from old epoch. forwarding boundary finalization"
+                    );
+
+                    // Forward the finalization to the sender. This operation is best-effort.
+                    let message = Voter::<Scheme<V>, H::Digest>::Finalization(finalization);
+                    if let Err(err) = recovered_global_sender.send(
+                        epoch,
+                        Recipients::One(from),
+                        message.encode().freeze(),
+                        true
+                    ).await {
+                        error!(?err, "failed to forward boundary finalization to peer");
+                    }
+                },
+                transition = self.mailbox.next() => {
+                    let Some(transition) = transition else {
+                        warn!("mailbox closed, shutting down orchestrator");
+                        break;
+                    };
 
-                    info!(transition.epoch, "entered new epoch");
-                }
-                Message::Exit(epoch) => {
-                    // Remove all entries with key less than or equal to the requested exit epoch.
-                    let epochs_to_remove: Vec<_> = engines
-                        .keys()
-                        .take_while(|k| **k <= epoch)
-                        .copied()
-                        .collect();
+                    match transition {
+                        Message::Enter(transition) => {
+                            // If the epoch is already in the map, ignore.
+                            if engines.contains_key(&transition.epoch) {
```

Comment:
Considered keeping a `min_epoch` here that was set when `exit` was called. Ultimately decided keeping it up to the caller was best (although we do a cleanup of anything less than `Exit()`.

Reply:
Going to align enter/exit behavior in a follow-up (unrelated to these changes).